app/routes/_auth.signup.tsx

```diff
@@ -14,32 +17,50 @@ export async function action({ request }: ActionArgs) {
   const password = form.get('password');
   const confirmPassword = form.get('confirmPassword');
 
-  if (!name || !email || !password || !confirmPassword) {
-    return { status: 400, message: 'Missing required fields' };
-  }
-
-  if (password !== confirmPassword) {
-    return { status: 400, message: 'Passwords do not match' };
-  }
+  const SignUpForm = z.object({
+    name: z
+      .string()
+      .trim()
+      .min(1, 'Name is required.')
+      .max(191, 'Name cannot exceed 191 max length.')
+      .transform((name) => name.replace(/\s+/g, ' ')),
+    email: z
+      .string()
+      .toLowerCase()
+      .trim()
+      .email('Invalid email.')
+      .refine(
+        async (email) => !(await db.user.findUnique({ where: { email } })),
+        'Email already in use. Is this you? Please log in instead.',
+      ),
+    password: z
+      .string()
+      .min(8, 'Password must be 8 or more characters.')
+      .max(191, 'Password cannot exceed 191 max length.'),
+    confirmPassword: z.string().refine((confirmPassword) => confirmPassword === password, 'Password must match.'),
+  });
 
-  const cleanEmail: string = email.toString().toLowerCase().trim();
-  const user = await db.user.findUnique({ where: { email: cleanEmail } });
+  const signUpForm = await SignUpForm.safeParseAsync({ name, email, password, confirmPassword });
 
-  if (user) {
-    return { status: 400, message: 'Email already in use. Is this you? Please log in instead.' };
+  if (!signUpForm.success) {
+    return badRequest({
+      success: false,
+      fields: { name, email },
```

Comment:
We are only returning `name` and `email` here and not the user `password` for security purposes.